### ISSUE TITLE
Long-term feature request: numeric chords

### ISSUE THREAD

**Issue by wolftune** [sciurius/chordpro#6]
My long-term ideal for ChordPro would be to support both classical (and jazz) Roman numerals like [V7] or [iii] as well as [Nashville Number System](https://en.wikipedia.org/wiki/Nashville_number_system) like [5^7] or [3-].

My primary goal is actually to just be able to use those for chords and print out leadsheets that just specify numbers and not letters (let the reader use whatever key they want). I'm not necessarily asking that there be a mode to have the program automatically swap letters for number and vice-versa (basically do the analysis / interpretation for you), but that would be superb if that were possible.

**Comment by AnniePerson** [comment]
I would like to be able to use German chord names: ... A, Bb, H, C... instead of ... A, Bb, B, C ...
That could use the same technical solution as Nashville numbers requested by wolftune

**Comment by wolftune** [comment]
I thought German names took A B♭ B C and made them A B H C. There's no B♭ in German system. Anyway, it would be nice if we could prohibit anyone anywhere from using the stupid German system… but I suppose if people keep using it, ChordPro could support it even though… :P

**Comment by AnniePerson** [comment]
As I live in Germany and use scores from various sources I cannot avoid having a mix of naming systems. That's why I prefer A Bb H C because it cannot be misunderstood. This naming scheme is called "Scandinavian" by Finale and "German" by MuseScore. A B H C is the traditional German system that is often found in classical music, Finale calls it "German", MuseScore calls it "Full German"

**Comment by sciurius** [comment]
The attached graphic shows experimental support for NNS.

If I could attach a graphic, that is :( .
See http://www.squirrel.nl/pub/xfer/uploads/3CWNuLlvwjS6z1fbHfQX8FlQ.png

**Comment by wolftune** [comment]
Holy crap, That's awesome!

Now that I see it, I realize I probably personally would rather [5m^7] vs [5-^7] which is to say that I'm not really a NNS person, but I respect the value of deferring to standards. I'm more used to the Roman Numeral approach, but I've seen in my teaching that sometimes things like NNS are preferred by some people.

Anyway, having *any* numeric system like this would be SUPER amazingly awesome. My dream is to publish (under free/libre/open licensing, CC-BY-SA where feasible) song collections with that sort of thing. I wish this were the standard way to publish songs instead of the pick-a-key approach.

I'm *thrilled* to see experimental support so quickly!

**Comment by sciurius** [comment]
And roman numerals: http://www.squirrel.nl/pub/xfer/uploads/3Ci7Jz4a05SbH17zIFMd9YXA.png

**Comment by sciurius** [comment]
Note that I'm not a NNS nor Roman person, so your feedback is **very** valuable.

**Comment by wolftune** [comment]
I can totally provide any and all necessary feedback, thanks!!!!

it *probably* makes sense to allow `b7` for the markup rather than the unicode `♭7` but it should certainly be rendered with ♭

(side-note: I made that with the marvels of the Linux compose-key. I just hit compose, # b to get that flat symbol (and compose # # to get ♯ )

**Comment by sciurius** [comment]
The problem with rendering unicode specials is that I am not in control of the fonts that are used/available. The user is. So I can not be certain that there **is** a ♭ and ♯ to be shown...

**Comment by sciurius** [comment]
If you check out the "dev" branch you can start playing with the new features.

**Comment by wolftune** [comment]
There's a few variations that are inconsistent. When it's Am7♭5 in letters, it's marked in numeric analysis commonly as iiø7 (where the ø7 should be superscript), it's called "half-diminished". https://en.wikipedia.org/wiki/Half-diminished_seventh_chord has some details describing the various ways this is shown (as one example that has variation).

Re; unicode symbols, it makes sense to make it an option where the default is to use the unicode but some setting could swap to the plain # and b. I'd suggest leaving that option out and only using unicode and see if anyone ever complains. The unicode should be default for sure.

**Comment by wolftune** [comment]
I'll check the dev branch and report back, thanks!

**Comment by wolftune** [comment]
For reference, the ideal *eventual* system would be where the numerals could be in the markup but the rendering could be switched back to letters by specifying a key (and then after that, the harder ideal of having markup that used letters and render with numerals, which would require specifying which letter is the key, i.e. 1).

One more difficult long-term issue: songs that change key… for strict numeral stuff there needs to just be something like a spot that says "key change: new 1 = old ♭3" or something like that. (Probably should make these notes be separate issues)

**Comment by wolftune** [comment]
Upon some testing:

* bII rendered that way really looks bad compared to ♭II — the unicode symbol is really needed
* to indicate ii° we need the unicode for that too as a 0 or o put superscript doesn't look good
* it's a bit odd that I can mix NNS, Roman numerals, and regular chords, but that's just up to me to get it right
* blank chord grids at the end don't make sense to include

**Comment by sciurius** [comment]
Thanks for testing. Your feedback is appreciated.

For the sake of clarity, the NNS and Roman support is very experimental and several things are not covered yet. For example, it must generate an error when styles are mixed, transposition is requested, and adding chord diagrams must be suppressed.

My experience with other chord-typesetting tools (playtab, irealcvt) has taught me that the exact look and positioning of special symbols may vary from font to font. So while ii⁰ (super zero) would be the correct symbol, ii° (degree) may look nicer on some fonts. Would it be acceptable to enforce a particular font (e.g. Bravura) when using NNS and Roman?

**Comment by sciurius** [comment]
> When it's Am7♭5 in letters, it's marked in numeric analysis commonly as iiø7
I would like to avoid rewriting chords. If the user wants iiø7 she can write iiø7. Same applies to 'm' versus '-' for minor. It's up to the user.

**Comment by wolftune** [comment]
Agreed that the easiest for variations of symbols is to go with whatever the markup entered.

> while ii⁰ (super zero) would be the correct symbol

I'm not at all sure that's even true. I think degree sign might be more correct actually

> Would it be acceptable to enforce a particular font (e.g. Bravura) when using NNS and Roman?

I think good defaults are better than enforced fonts

**Comment by sciurius** [comment]
The chord naming 'bible' (Brandt, Carl & Roemer, Clinton - Standardized Chord Symbol Notation, 1976) calls it "a small circle". I must agree that the degree symbol (°) has a bigger chance to be circular than a zero or o . In hand-written music, most often a zero is used.
Here are some choices: ○ ° o ⚪ ⚬ ￮ .

> I think good defaults are better than enforced fonts

I agree fully. So no rewriting of chords, just print whatever the user passed in. If the user writes special symbols, she should designate a chords font that has glyphs for these symbols.

BTW: The only exception I made so far is to raise the chord quality, e.g., IV<sup>7</sup>, 4<sup>sus</sup>. I think this is desirable for Roman and necessary for Nashville. This doesn't require special fonts.

**Comment by wolftune** [comment]
> no rewriting of chords, just print whatever the user passed in

Well, I really think that it would make sense to translate (as an example) Ab to A♭ because it's likely that users don't know how to get the ♭ symbol but *do* want it. My suggestion is to make that translation of b -> ♭ and # -> ♯ as the default approach but have an argument that blocks that. Alternatively, if you want to stay on the safer side for fonts (although I suspect most common fonts *do* have those symbols), make the default not translate but *do* offer an argument that causes that translation to happen.

To be clear, I think specifying the font is not the only issue. Someone could specify a font that has the symbols but still be processing from markup that uses the regular b and # chars. So, it's not good to require people to enter the special character in every case in the markup in order to render with the special character.

> ○ ° o ⚪ ⚬ ￮

I agree that the 4th in that list looks best. I'd be curious to check in standard *music* fonts how the symbol is specified…

**Comment by sciurius** [comment]
Biggest obstacle is that there are few fonts out there that actually have a ♭ and ♯ ...
Do you know some?

**Comment by wolftune** [comment]
> there are few fonts out there that actually have a ♭ and ♯ 

That's not my experience at all. I'm barely finding any fonts that *don't* have them.

Whatever font we're using here on GitHub has them.

Some fonts on my machine that have them:

* Andale Mono
* Arial
* Deja Vu
* FreeMono, FreeSans, FreeSerif
* Georgia
* Hack
* Impact
* Liberation Serif
* Times New Roman
* Vera Sans

Okay, this is silly. I stopped checking. The ONLY font I found that *didn't* have them was webdings. Even the handwritten Nina font for Nina Paley's style a la [Mimi & Eunice](http://mimiandeunice.com) showed the symbols correctly, just not in Nina's style. It looks like almost every font out their either has its own style for these (I did see a number of variations among fonts) or just shares the same symbol from other fonts even though it doesn't really fit the look of the font in question.

So, I think it's actually safe to assume that these symbols are effectively *universally* available. They seem as reliable as © and ™ even. Do you know of *any* font that fails to show them?

**Comment by sciurius** [comment]
I'm surprised. All of these on my system do *not* have them.
I'll take a look again tomorrow.
Thanks for your valuable input.

**Comment by wolftune** [comment]
Well, interesting. To be clear: I just opened up LibreOffice and typed the symbols and then changed fonts… I think there's functions (not sure if this is just in LibreOffice) to fall-back to a standard symbol when it is missing… guessing…

**Comment by sciurius** [comment]
I was suspecting something like that.

**Comment by wolftune** [comment]
Of course, if that sort of fallback works fine within a PDF output, then we're fine

**Comment by sciurius** [comment]
Unfortunately, it doesn't... Apparently you need special libaries and/or coding.

I propose to release the NNS and Roman stuff in 'plain' form, without fancy substituting. This can be added later when there's reliable font support.

**Comment by sciurius** [comment]
I consider the current dev version mature w.r.t. Roman and NNS chords. I'd appreciate a good testing.

**Comment by wolftune** [comment]
I'm happy to have this go ahead into master. I have a couple other thoughts and think they should probably be separate issues for later (not block this merging of current functionality). I've tested it and confirm that the core functions are working consistently enough for a first release to master of the functionality.

**Comment by wolftune** [comment]
Okay, I did find a couple issues:

* [3ø] should render the ø in superscript but it doesn't.
* [4Δ] rendered as 4¿
    * [4maj7] works somewhat though it's not real NNS, not sure if the "maj7" should be superscript (it's not right now)

**Comment by sciurius** [comment]
I have added support for ♯ ♭ Δ ° and ø in Nashville/Roman chords. Note that there is no substitution done, they are just recognized in the input and passed to the output. The font used for chords must have these glyphs or you will get tofu.
I've also added a sample config for songs with Nashville/Roman chord names.

chordpro --config nashville ...

This will use the default config, and override the font for chords by a serif font (Times-Roman).
